### PR TITLE
Bugfix FXIOS-13855 [Share Manager] Add remoteURL capability to .file share types for the Send to Device activity

### PR DESCRIPTION
**Pull request by ih-codes**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-13855)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/30012)

#### New:
This PR addresses the issue in FXIOS-13855 where online PDFs viewed in a tab did not show "Send Link to Device" as a share option in the share sheet. This was by original design, but the code has been updated now to include a "remote URL" for `.file` type shares. This means online PDFs viewed in a tab can now be shared via this remote URL to other Firefox synced devices, like your desktop browser.

Note that files shared from the Downloads Panel, or opened into a tab from the Downloads Panel, do _not_ have a remote URL. That means "Send Link to Device" will not appear in the share options.

(Testing steps in ticket under QA notes)

##### Demo:

https://github.com/user-attachments/assets/0be65fa0-17d2-4492-bc26-431c4b2e4166

#### Old:
Just a quick draft reference PR for discussion on #30059.

## :pencil: Checklist
- [x] I filled in the ticket numbers and a description of my work
- [x] I updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] I ensured unit tests pass and wrote tests for new code
- [x] If working on UI, I checked and implemented accessibility (Dynamic Text and VoiceOver)
- [ ] If adding telemetry, I read the [data stewardship requirements](https://github.com/mozilla-mobile/firefox-ios/wiki/Adding-Glean-Telemetry-Events) and will request a data review
- [ ] If adding or modifying strings, I read the [guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/How-to-add-and-modify-Strings) and will request a string review from l10n
- [ ] If needed, I updated documentation and added comments to complex code

### CODE REVIEW

**Review by dicarobinho — APPROVED** [review]
LGTM!

**Comment by ih-codes** [comment]
Thanks @dicarobinho for testing and updating the unit tests! 🎉  Opening this PR now.

File: firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift [hunk 1]
```diff
@@ -680,9 +680,12 @@ class BrowserCoordinator: BaseCoordinator,
 
     func presentSavePDFController() {
         guard let selectedTab = browserViewController.tabManager.selectedTab else { return }
+
         if selectedTab.mimeType == MIMEType.PDF {
             showShareSheetForCurrentlySelectedTab()
         } else {
+            let remoteURL = selectedTab.webView?.url
```
Comment (cyndichin) [inline firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift:688]: For context, maybe we can add a comment on why we have this remoteURL. I liked how you explained it here in the PR description.

`This means online PDFs viewed in a tab can now be shared via this remote URL to other Firefox synced devices, like your desktop browser.`

I see you mention more in `ShareType`, but maybe helpful here too? Or expand over there?

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/Sharing/ShareManagerTests.swift [hunk 2]
```diff
@@ -36,7 +36,7 @@ final class ShareManagerTests: XCTestCase {
         let testShareActivityType = UIActivity.ActivityType.message
 
         let activityItems = ShareManager.getActivityItems(
-            forShareType: .file(url: testFileURL),
+            forShareType: .file(url: testFileURL, remoteURL: nil),
```
Comment (cyndichin) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/Sharing/ShareManagerTests.swift:39]: Do we want to add tests for when remoteURL != nil.

File: firefox-ios/Client/Frontend/Browser/MainMenuActionHelper.swift [hunk 3]
```diff
@@ -669,10 +670,10 @@ final class MainMenuActionHelper: PhotonActionSheetProtocol,
     /// non-HTML MIME type currently opened in the webView, such as a PDF).
     /// NOTE: Called from getShareFileAction (files in the browser) and getShareAction (websites)
     @MainActor
-    private func share(fileURL: URL, buttonView: UIView) {
+    private func share(fileURL: URL, remoteURL: URL?, buttonView: UIView) {
         TelemetryWrapper.recordEvent(category: .action, method: .tap, object: .sharePageWith)
         navigationHandler?.showShareSheet(
-            shareType: .file(url: fileURL),
+            shareType: .file(url: fileURL, remoteURL: remoteURL),
```
Comment (cyndichin) [inline firefox-ios/Client/Frontend/Browser/MainMenuActionHelper.swift]: instead of passing in `nil`, could we set `remoteURL: nil` and add the comment here instead. Unless there is an existing case in which we want the value to not be nil?

File: firefox-ios/Client/Coordinators/ShareSheetCoordinator.swift [hunk 4]
```diff
@@ -90,17 +90,26 @@ class ShareSheetCoordinator: BaseCoordinator,
...
-            // Cannot send file:// URLs to another synced device
-            guard !shareType.wrappedURL.isFileURL else {
+            var sendURL: URL = shareType.wrappedURL
+
+            if case .file(_, let remoteURL) = shareType,
+               let remoteURL = remoteURL {
+                // Some files might have a remote URL as a fallback for Send to Device (i.e. PDFs navigated to in a tab, but
+                // not from Downloads Panel).
+                sendURL = remoteURL
+            }
```
Comment (cyndichin) [inline firefox-ios/Client/Coordinators/ShareSheetCoordinator.swift:100]: would it be cleaner if we add this to `ShareType` so that we can access `shareType.remoteURL` . No strong opinions, but seems more readable.

File: firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift [hunk 5]
```diff
@@ -1337,7 +1340,7 @@ class BrowserCoordinator: BaseCoordinator,
         }
 
         // If we successfully got a temp file URL, share it like a downloaded file
-        return .file(url: fileURL)
+        return .file(url: fileURL, remoteURL: shareType.wrappedURL)
```
Comment (cyndichin) [inline firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift]: I was a bit confused at first since the wrappedURL is pulling from the `url` and not `remoteURL` in `ShareType`, but it seems this is in the right format so we use the existing shareType url for the new shared type.

File: firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift (hunk 1 above)
Comment (ih-codes) [reply firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift:688]: Sure I can include an extra comment here for context! 👍

File: firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift (hunk 5 above)
Comment (ih-codes) [reply firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift]: Edit: Replied to the wrong comment, let me move what I'd said to the right chain!

---

Yeahhhhh I was a bit concerned the 2 URLs might be confusing. I am open to adjusting the naming or documentation anywhere I can to make this clearer. 🙏 

The distinction here is that the wrapped file URL is always a `file://` URL, whereas the remoteURL will always (if present) link to online web content (a `https://` type URL).

File: firefox-ios/Client/Coordinators/ShareSheetCoordinator.swift (hunk 4 above)
Comment (ih-codes) [reply firefox-ios/Client/Coordinators/ShareSheetCoordinator.swift:100]: Every share type has a default URL for content on the device itself (`file://` for `file` shares, `https://` for `tab` and `site` shares). Only the `file` case has a special secondary URL, so I don't think it's appropriate to add a wrapper on the entire enum. It's the same as how `tab` also has a `any ShareTab` but it doesn't really make sense to have a `wrappedTab` accessed from a `file` share, IMO.

File: firefox-ios/Client/Frontend/Browser/MainMenuActionHelper.swift (hunk 3 above)
Comment (ih-codes) [reply firefox-ios/Client/Frontend/Browser/MainMenuActionHelper.swift]: Ahh good catch, I'll remove the param and just set `nil` here with a comment now!

File: firefox-ios/Client/Coordinators/ShareSheetCoordinator.swift (hunk 4 above)
Comment (cyndichin) [reply firefox-ios/Client/Coordinators/ShareSheetCoordinator.swift:100]: Hmm, good point! Okay this works for me then!

File: firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift (hunk 5 above)
Comment (cyndichin) [reply firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift]: If you could elaborate why we are passing in the wrapped file in the remoteURL that would be helpful.

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/Sharing/ShareManagerTests.swift (hunk 2 above)
Comment (ih-codes) [reply firefox-ios/firefox-ios-tests/Tests/ClientTests/Sharing/ShareManagerTests.swift:39]: Good call, I'll add this. Also, I'll look at adding some `SendToDeviceActivity` tests since I noticed we don't have any specific to custom app activities. It's a little different than the other share types.

File: firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift (hunk 5 above)
Comment (ih-codes) [reply firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift]: So, `tryDownloadingTabFileToShare` will download the PDF in the background when a user views an online PDF in the browser. Then, if the user tries to share the current tab, the share activity gets the file to attach to emails, iMessages, etc. But for Send to Device, we actually just want to send the link, not the downloaded file. So the shareType passed into `tryDownloadingTabFileToShare` is always a tab share and thus we can get the "remote URL" from the tab share's wrapped URL. I will add some comments to hopefully make this a little clearer.

Comment (cyndichin) [reply firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift]: Ahhh okay that makes sense. Yeah I think if you add that, it will be a lot more helpful. Maybe adding this would help too? With a clarifying comment above?
`let tabRemoteUrl = shareType.wrappedURL`

Comment (ih-codes) [reply firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift]: Actually, I just finished restructuring this function and adding some comments. I think you'll find it a lot clearer now since it can only apply the logic to a tabURL and a ShareTabe instead of a generic ShareType. Will commit soon after I test! 😁

Comment (ih-codes) [reply firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift]: See if you think it's clearer now @cyndichin !

**Comment by ih-codes** [comment]
@cyndichin I've addressed all your comments and added a bunch more tests!

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/Sharing/ShareManagerTests.swift [hunk 6]
```diff
@@ -32,11 +32,62 @@ final class ShareManagerTests: XCTestCase {
...
+            createStubActivityViewController(),
+            itemForActivityType: testShareActivityType
+        )
+
+        XCTAssertEqual(activityItems.count, 2)
+        XCTAssertEqual(itemForURLActivity as? URL, testFileURL)
+        XCTAssertTrue(itemForShareActivity is NSNull)
+    }
+
+    func testGetActivityItems_forFileURL_withRemoteURL_withNoShareText() throws {
```
Comment (cyndichin) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/Sharing/ShareManagerTests.swift:60]: duplicate test?

File: firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift (hunk 5 above)
Comment (cyndichin) [reply firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift]: Oh yes much better!!! Thank you for restructuring it! 🥳

**Comment by ih-codes** [comment]
> Looks good to me now, way better!! Thank you!!! Seems like there may be a dupe in the tests? If it passes CI, then good to merge! Don't want to block ya since its towards my EOD.

Yep whoops, just noticed the dupe I pushed up before my meeting. It's fixed now! 😆  Always fun writing tests for really old code (Send to Device) lol. Thanks for the thorough review! 🙏
